### DISCUSSION TITLE
Question regarding roles & security

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#4409] [Archive]
Hello, first of all thank you for Directus.

I'm using the GraphQL endpoint with a custom Role. The problem is that the role can list directus users, permissions and more as long as the "App Access" is checked. Unchecking App Access returns a 403 for all requests.

So :

**Scenario 1**
Role: Custom
Permissions: Read all bookings
App Access: Checked
Admin: NOT Checked
Result: User can access the bookings, but can also access directus_users, permissions... and list them (security problem)

**Scenario 2**
Role: Custom
Permissions: Read all bookings
App Access: NOT Checked
Admin: NOT Checked
Result: User CAN'T access the bookings, and CAN'T access directus_users, permissions...

What I think should be is

**Scenario 3**
Role: Custom
Permissions: Read all bookings
App Access: ?
Admin: NOT Checked
Result: User can access the bookings, but CAN'T access directus_users, permissions... and list them (security problem)

Also, may be App Access should be more clear? Is it App Access or Api Access ? What about a need for a role that should only use API and not access the App ?

**Comment by ghost** [comment]
I've just seen "App Access Minimum" button which answers some of my questions above, shouldn't it be by default ? Because I discovered the issues above by chance. Also, even with App Access Minimum, the custom role can list all collections & fields, settings..., which also can be a security problem.

Is it possible to have a custom role that, if configured, obeys permissions strictly without any access to directus_* data ?

What do you think? Thanks.

**Comment by ghost** [comment]
Also, in permissions, and using REST calls:

Scenario 1, enabling permission
Role: Custom
Permissions: Read all bookings
App Access: Checked
Admin: NOT Checked
Result: User can access the bookings

Scenario 2, disabling the permission above
Role: Custom
Permissions: CAN'T read bookings at all
App Access: Checked
Admin: NOT Checked
Result: User can still access the bookings.

It's like disabling a permission, after having enabled it, is not taken into consideration.

**Comment by rijkvanzanten** [comment]
> Is it possible to have a custom role that, if configured, obeys permissions strictly without any access to directus_* data ?

Yes, this is the difference between app and no-app access. App Access means as much as "Do the users in this role have access to use the Directus app". The Directus app relies on the information coming from fields/collections/permissions and some others to render out the app, so those would then also be accessible through the API.

**Reply by rijkvanzanten** [reply]
(As for the security aspect, they'd only be able to read the information on fields/collections/permissions that applies to their particular role, not all of it)
